File: week6/Adapter/Adapter.cpp
#include <iostream>
#include <string>

class InternalClass {
public:
    std::string fetch() {
        return "get user info";
    }
};

class ExternalClass {
public:
    std::string search() {
        return "get user info";
    }
};

class Adapter {
private:
    ExternalClass* external;
public:
    Adapter(ExternalClass* ext) : external(ext) {}

    std::string fetch() {
        return external->search();
    }
};

int main() {
    InternalClass internal;
    std::cout << "Internal: " << internal.fetch() << std::endl;

    ExternalClass external;
    Adapter adapter(&external);
    std::cout << "External through Adapter: " << adapter.fetch() << std::endl;

    return 0;
}

File: week6/Adapter/Adapter.kt
class InternalClass {
    fun fetch(): String {
        return "get user info"
    }
}

class ExternalClass {
    fun search(): String {
        return "get user info"
    }
}

class Adapter(private val external: ExternalClass) {
    fun fetch(): String {
        return external.search()
    }
}

fun main() {
    val internal = InternalClass()
    println("Internal: ${internal.fetch()}")

    val external = ExternalClass()
    val adapter = Adapter(external)
    println("External through Adapter: ${adapter.fetch()}")
}
